src/Swarm/Game/Scenario/WorldDescription.hs

```diff
@@ -53,9 +63,108 @@ instance FromJSONE (EntityMap, RobotMap) WorldDescription where
 --   character in the palette, failing if any character in the raw map
 --   is not contained in the palette.
-paintMap :: MonadFail m => WorldPalette -> Text -> m [[Cell]]
+paintMap :: MonadFail m => WorldPalette Cell -> Text -> m [[Cell]]
 paintMap pal = traverse (traverse toCell . into @String) . T.lines
  where
   toCell c = case KeyMap.lookup (Key.fromString [c]) (unPalette pal) of
     Nothing -> fail $ "Char not in world palette: " ++ show c
     Just cell -> return cell
+
+------------------------------------------------------------
+-- World editor
+------------------------------------------------------------
+
+type TerrainEntityNamePair = (TerrainType, Maybe EntityName)
+
+type TerrainEntityFacadePair = (TerrainType, Maybe EntityFacadeData)
+
+cellToTerrainEntityNamePair :: CellPaintDisplay -> TerrainEntityFacadePair
+cellToTerrainEntityNamePair (Cell terrain maybeEntity _) = (terrain, maybeEntity)
+
+toCellPaintDisplay :: Cell -> CellPaintDisplay
+toCellPaintDisplay (Cell terrain maybeEntity r) =
+  Cell terrain (mkPaint <$> maybeEntity) r
+
+type WorldDescriptionPaint = ProtoWorldDescription CellPaintDisplay
+
+instance ToJSON WorldDescriptionPaint where
+  toJSON w =
+    object
+      [ "default" .= defaultTerrain w
+      , "offset" .= offsetOrigin w
+      , "palette" .= Y.toJSON paletteKeymap
+      , "upperleft" .= ul w
+      , "map" .= Y.toJSON mapText
+      ]
+   where
+    cellGrid = area w
+    suggestedPalette = palette w
+    (mapText, paletteKeymap) = prepForJson suggestedPalette cellGrid
+
+toKey :: TerrainEntityFacadePair -> TerrainEntityNamePair
+toKey = fmap $ fmap (\(EntityFacadeData eName _display) -> eName)
+
+getUniquePairs :: [[CellPaintDisplay]] -> M.Map TerrainEntityNamePair TerrainEntityFacadePair
+getUniquePairs cellGrid =
+  M.fromList $ concatMap (map genTuple) cellGrid
+ where
+  genTuple c =
+    (toKey terrainEfd, terrainEfd)
+   where
+    terrainEfd = cellToTerrainEntityNamePair c
+
+-- | Note that display characters are not unique
+-- across different entities! However, the palette KeyMap
+-- as a conveyance serves to dedupe them.
+prepForJson ::
+  WorldPalette CellPaintDisplay ->
+  [[CellPaintDisplay]] ->
+  (Text, KM.KeyMap CellPaintDisplay)
+prepForJson (WorldPalette suggestedPalette) cellGrid =
+  (renderFullMap cellGrid, paletteKeymap)
+ where
+  preassignments :: [(Char, TerrainEntityFacadePair)]
+  preassignments =
+    map (first T.head . fmap cellToTerrainEntityNamePair) $
+      M.toList $
+        KM.toMapText suggestedPalette
+
+  entityCells :: M.Map TerrainEntityNamePair TerrainEntityFacadePair
+  entityCells = getUniquePairs cellGrid
+
+  unassignedCells :: M.Map TerrainEntityNamePair TerrainEntityFacadePair
+  unassignedCells =
+    M.withoutKeys entityCells $
+      Set.fromList $
+        map (toKey . snd) preassignments
+
+  unassignedCharacters :: Set.Set Char
+  unassignedCharacters =
+    -- TODO: Make this pool larger?
+    Set.difference (Set.fromList ['a' .. 'z']) $
```

Comment:
I would suggest filtering all the `Char` values based on their Unicode category.

Letters, numbers, symbols and marks should give us a big enough pool. 🙂

Reply:
Yes, I was thinking about that.  And `Data.Char` has functions that allow filtering by category.  But what should the code look like that generates the initial list to be filtered?